tools/validate_spec.py

```diff
@@ -79,81 +77,106 @@ def cli(argv: List[str]) -> Union[int, str]:
-            for message_file_path in path_generator:
-                msg = get_message_to_validate(message_file_path, path == path_valid)
-                validation_result = validator.validate_message(msg)
-                json_output_writer.write(validation_result)
-                if validation_result.classification in [
-                    Classification.FP,
-                    Classification.FN,
-                ]:
-                    print(f"{message_file_path}: FAILED")
-                    print(f"{validation_result.get_abbreviated_output()}\n")
-                    if abort_on_error:
-                        raise ValidationError(
-                            f"{message_file_path} classified as "
-                            f"{validation_result.classification.value}"
-                        )
-                    classified_incorrectly += 1
-                else:
-                    print(f"{message_file_path}: PASSED")
-
-    if classified_incorrectly != 0:
-        raise ValidationError(f"{classified_incorrectly} messages were classified incorrectly")
-
-
-class JsonOutputWriter:
+    def validate(original_message: bytes, original_message_is_valid: bool, msg_path: Path) -> bool:
+        parser_result = rflx_parse_message_from_bytes(original_message)
+        parsed_message_is_valid = False
+
+        if isinstance(parser_result, MessageValue):
+            parsed_message_is_valid = (
+                parser_result.bytestring == original_message and parser_result.valid_message
+            )
+
+        if original_message_is_valid:
+            validation_success = parsed_message_is_valid
+        else:
+            validation_success = not parsed_message_is_valid
+
+        if validation_success:
+            output_writer.passed(
+                parser_result,
+                msg_path,
+                original_message,
+                original_message_is_valid,
+                parsed_message_is_valid,
+            )
+        else:
+            output_writer.failed(
+                parser_result,
+                msg_path,
+                original_message,
+                original_message_is_valid,
+                parsed_message_is_valid,
+            )
+
+        return validation_success
+
+    valid_message_paths: Optional[Iterator[Path]] = (
+        path_valid.glob("*") if path_valid is not None else path_valid
+    )
+    invalid_message_paths: Optional[Iterator[Path]] = (
+        path_invalid.glob("*") if path_invalid is not None else path_invalid
+    )
+
+    with OutputWriter(full_output_path) as output_writer:
+        for paths in [valid_message_paths, invalid_message_paths]:
+            if paths is None:
+                continue
+            for path in paths:
+                original_message = get_message_from_path(path)
+                val_success = validate(original_message, paths == valid_message_paths, path)
+                if not val_success and abort_on_error:
+                    return
```

Comment:
Long inner functions make it hard to see what the actual code this function is. Changing the inner functions into top-level private functions would make the code cleaner and more readable.

Reply:
There are now only two short inner functions left